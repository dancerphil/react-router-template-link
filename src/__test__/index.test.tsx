import * as React from 'react';
import '@testing-library/jest-dom';
import {create} from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom';
import createFactory from '..';

const {createLink, Link} = createFactory();

describe('createFactory', () => {
    test('createLink', () => {
        const HomeLink = createLink('/');
        const link = <HomeLink>text</HomeLink>;
        expect(create(link)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link}</BrowserRouter>)).toMatchSnapshot();
    });

    test('Link', () => {
        expect(create(<Link to="/">text</Link>)).toMatchSnapshot();
    });

    test('template', () => {
        interface Params {
            userId: string;
        }
        const UserLink = createLink<Params>('/users/{userId}');
        const link = <UserLink userId="danceprhil">text</UserLink>;
        expect(create(link)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link}</BrowserRouter>)).toMatchSnapshot();
    });

    test('query', () => {
        interface Params {
            userId: string;
            from: string;
        }
        const UserLink = createLink<Params>('/users/{userId}');
        const link = <UserLink userId="danceprhil" from="file">text</UserLink>;
        expect(create(link)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link}</BrowserRouter>)).toMatchSnapshot();
    });

    test('explicit query', () => {
        interface Params {
            userId: string;
            from: string;
        }
        const UserLink = createLink<Params>('/users/{userId}?from={from}');
        const link = <UserLink userId="danceprhil" from="file">text</UserLink>;
        expect(create(link)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link}</BrowserRouter>)).toMatchSnapshot();
    });

    test('combined query', () => {
        interface Params {
            userId: string;
            from: string;
            type: string;
        }
        const UserLink = createLink<Params>('/users/{userId}?from={from}');
        const url = UserLink.toUrl({userId: 'danceprhil', from: 'file', type: 'a'});
        const link = <UserLink userId="danceprhil" from="file" type="a">text</UserLink>;
        expect(url).toBe('/users/danceprhil?from=file&type=a');
        expect(create(link)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link}</BrowserRouter>)).toMatchSnapshot();
    });

    test('encodePathVariable', () => {
        const {createLink} = createFactory({encodePathVariable: true});
        interface Params {
            userId: string;
        }
        const UserLink = createLink<Params>('/users/{userId}');
        const link = <UserLink userId="a/b">text</UserLink>;
        expect(create(link)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link}</BrowserRouter>)).toMatchSnapshot();
    });

    test('hash', () => {
        const HomeLink = createLink('/');
        const link = <HomeLink hash="1">text</HomeLink>;
        expect(create(link)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link}</BrowserRouter>)).toMatchSnapshot();
    });

    test('external', () => {
        const ExampleLink = createLink('https://example.com');
        const link = <ExampleLink>text</ExampleLink>;
        expect(create(link)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link}</BrowserRouter>)).toMatchSnapshot();
    });

    test('react-router-dom className props', () => {
        const HomeLink = createLink('/');
        const link1 = <HomeLink className="class1">text</HomeLink>;
        const link2 = <HomeLink className={() => 'class2'}>text</HomeLink>;
        // @ts-expect-error
        const link3 = <HomeLink className={{}}>text</HomeLink>;
        expect(create(link1)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link1}</BrowserRouter>)).toMatchSnapshot();
        expect(create(link2)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link2}</BrowserRouter>)).toMatchSnapshot();
        expect(create(link3)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link3}</BrowserRouter>)).toMatchSnapshot();
    });

    test('react-router-dom style props', () => {
        const HomeLink = createLink('/');
        const link1 = <HomeLink style={{color: 'pink'}}>text</HomeLink>;
        const link2 = <HomeLink style={() => ({color: 'hotpink'})}>text</HomeLink>;
        expect(create(link1)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link1}</BrowserRouter>)).toMatchSnapshot();
        expect(create(link2)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link2}</BrowserRouter>)).toMatchSnapshot();
    });

    test('react-router-dom children props', () => {
        const HomeLink = createLink('/');
        const link1 = <HomeLink>{() => 'text'}</HomeLink>;
        // @ts-expect-error
        const link2 = <HomeLink />;
        expect(create(link1)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link1}</BrowserRouter>)).toMatchSnapshot();
        expect(create(link2)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link2}</BrowserRouter>)).toMatchSnapshot();
    });

    test('react-router-dom to props', () => {
        // @ts-expect-error
        const link1 = <Link to={undefined}>text</Link>;
        const link2 = <Link to={{pathname: '/'}}>text</Link>;
        expect(create(link1)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link1}</BrowserRouter>)).toMatchSnapshot();
        expect(create(link2)).toMatchSnapshot();
        expect(create(<BrowserRouter>{link2}</BrowserRouter>)).toMatchSnapshot();
    });
});
