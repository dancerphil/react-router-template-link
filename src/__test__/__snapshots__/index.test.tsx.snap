// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createFactory Link 1`] = `
<a
  className="panda-link panda-link-default"
  href="/"
>
  text
</a>
`;

exports[`createFactory combined query 1`] = `
<a
  className="panda-link panda-link-default"
  href="/users/danceprhil?from=file&type=a"
>
  text
</a>
`;

exports[`createFactory combined query 2`] = `
<a
  className="panda-link panda-link-default"
  href="/users/danceprhil?from=file&type=a"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory createLink 1`] = `
<a
  className="panda-link panda-link-default"
  href="/"
>
  text
</a>
`;

exports[`createFactory createLink 2`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default active"
  href="/"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory encodePathVariable 1`] = `
<a
  className="panda-link panda-link-default"
  href="/users/a%2Fb"
>
  text
</a>
`;

exports[`createFactory encodePathVariable 2`] = `
<a
  className="panda-link panda-link-default"
  href="/users/a%2Fb"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory explicit query 1`] = `
<a
  className="panda-link panda-link-default"
  href="/users/danceprhil?from=file"
>
  text
</a>
`;

exports[`createFactory explicit query 2`] = `
<a
  className="panda-link panda-link-default"
  href="/users/danceprhil?from=file"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory external 1`] = `
<a
  className="panda-link panda-link-default"
  href="https://example.com"
  rel="noopener noreferrer"
  target="_blank"
>
  text
</a>
`;

exports[`createFactory external 2`] = `
<a
  className="panda-link panda-link-default"
  href="https://example.com"
  rel="noopener noreferrer"
  target="_blank"
>
  text
</a>
`;

exports[`createFactory hash 1`] = `
<a
  className="panda-link panda-link-default"
  href="/#1"
>
  text
</a>
`;

exports[`createFactory hash 2`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default active"
  href="/#1"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory query 1`] = `
<a
  className="panda-link panda-link-default"
  href="/users/danceprhil?from=file"
>
  text
</a>
`;

exports[`createFactory query 2`] = `
<a
  className="panda-link panda-link-default"
  href="/users/danceprhil?from=file"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory react-router-dom children props 1`] = `
<a
  className="panda-link panda-link-default"
  href="/"
>
  text
</a>
`;

exports[`createFactory react-router-dom children props 2`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default active"
  href="/"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory react-router-dom children props 3`] = `
<a
  className="panda-link panda-link-default"
  href="/"
/>
`;

exports[`createFactory react-router-dom children props 4`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default active"
  href="/"
  onClick={[Function]}
/>
`;

exports[`createFactory react-router-dom className props 1`] = `
<a
  className="panda-link panda-link-default class1"
  href="/"
>
  text
</a>
`;

exports[`createFactory react-router-dom className props 2`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default class1 active"
  href="/"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory react-router-dom className props 3`] = `
<a
  className="panda-link panda-link-default class2"
  href="/"
>
  text
</a>
`;

exports[`createFactory react-router-dom className props 4`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default class2"
  href="/"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory react-router-dom className props 5`] = `
<a
  className="panda-link panda-link-default"
  href="/"
>
  text
</a>
`;

exports[`createFactory react-router-dom className props 6`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default active"
  href="/"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory react-router-dom style props 1`] = `
<a
  className="panda-link panda-link-default"
  href="/"
  style={
    Object {
      "color": "pink",
    }
  }
>
  text
</a>
`;

exports[`createFactory react-router-dom style props 2`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default active"
  href="/"
  onClick={[Function]}
  style={
    Object {
      "color": "pink",
    }
  }
>
  text
</a>
`;

exports[`createFactory react-router-dom style props 3`] = `
<a
  className="panda-link panda-link-default"
  href="/"
  style={
    Object {
      "color": "hotpink",
    }
  }
>
  text
</a>
`;

exports[`createFactory react-router-dom style props 4`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default active"
  href="/"
  onClick={[Function]}
  style={
    Object {
      "color": "hotpink",
    }
  }
>
  text
</a>
`;

exports[`createFactory react-router-dom to props 1`] = `
<a
  className="panda-link panda-link-default"
  href=""
>
  text
</a>
`;

exports[`createFactory react-router-dom to props 2`] = `
<a
  className="panda-link panda-link-default"
  href=""
>
  text
</a>
`;

exports[`createFactory react-router-dom to props 3`] = `
<a
  className="panda-link panda-link-default"
  href="/"
>
  text
</a>
`;

exports[`createFactory react-router-dom to props 4`] = `
<a
  aria-current="page"
  className="panda-link panda-link-default active"
  href="/"
  onClick={[Function]}
>
  text
</a>
`;

exports[`createFactory template 1`] = `
<a
  className="panda-link panda-link-default"
  href="/users/danceprhil"
>
  text
</a>
`;

exports[`createFactory template 2`] = `
<a
  className="panda-link panda-link-default"
  href="/users/danceprhil"
  onClick={[Function]}
>
  text
</a>
`;
